org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corrections/proposals/GradleCompatibilityProcessor.java

```diff
@@ -0,0 +1,104 @@
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+package org.eclipse.jdt.ls.core.internal.corrections.proposals;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.Collection;
+
+import org.eclipse.buildship.core.internal.CorePlugin;
+import org.eclipse.buildship.core.internal.preferences.PersistentModel;
+import org.eclipse.buildship.core.internal.util.gradle.GradleVersion;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IModuleDescription;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
+import org.eclipse.jdt.internal.core.JarPackageFragmentRoot;
+import org.eclipse.jdt.internal.ui.text.correction.IProblemLocationCore;
+import org.eclipse.jdt.ls.core.internal.ProjectUtils;
+import org.eclipse.jdt.ls.core.internal.commands.ProjectCommand;
+import org.eclipse.jdt.ls.core.internal.commands.ProjectCommand.ClasspathResult;
+import org.eclipse.jdt.ls.core.internal.corrections.CorrectionMessages;
+import org.eclipse.jdt.ls.core.internal.corrections.IInvocationContext;
+import org.eclipse.jdt.ls.core.internal.managers.GradleUtils;
+import org.eclipse.jdt.ls.core.internal.text.correction.CUCorrectionCommandProposal;
+import org.eclipse.lsp4j.CodeActionKind;
+
+public class GradleCompatibilityProcessor {
+	public static void getGradleCompatibilityProposals(IInvocationContext context, IProblemLocationCore problem, Collection<ChangeCorrectionProposal> proposals) {
+		IJavaProject javaProject = context.getCompilationUnit().getJavaProject();
+		if (javaProject == null) {
+			return;
+		}
+		IProject project = javaProject.getProject();
+		if (!ProjectUtils.isGradleProject(project)) {
+			return;
+		}
+		PersistentModel model = CorePlugin.modelPersistence().loadModel(project);
+		if (!model.isPresent()) {
+			return;
+		}
+		GradleVersion gradleVersion = model.getGradleVersion();
+		if (gradleVersion != null && gradleVersion.compareTo(GradleVersion.version(GradleUtils.JPMS_SUPPORTED_VERSION)) < 0) {
+			IResource resource = javaProject.getResource();
+			if (resource == null) {
+				return;
+			}
+			URI uri = resource.getLocationURI();
+			if (uri == null) {
+				return;
+			}
+			try {
+				ClasspathResult result = ProjectCommand.getClasspathsFromJavaProject(javaProject, new ProjectCommand.ClasspathOptions());
+				IModuleDescription moduleDescription = javaProject.getModuleDescription();
+				if (moduleDescription == null) {
```

Comment:
Suggest to extract the code in if-else blocks to methods separately - for better readiness.

Reply:
done in https://github.com/eclipse/eclipse.jdt.ls/pull/2304/commits/daf2a764ac1bf102b523b95b6e99648fe7663ad9